### PR TITLE
Issue #2796005 by mglaman: Chain resolvers do not support tagged serv…

### PR DESCRIPTION
**Pull request by mglaman**
…ice priority

### CODE REVIEW

**Comment by bojanz** [comment]
Defaulting priority to 0 means that any call to addResolver() that doesn't have a priority specified will overwrite the first resolver in the list.

**Comment by mglaman** [comment]
That's what I get for copying core.

**Comment by mglaman** [comment]
@bojanz updated, removed default parameter value for $priority.

File: modules/order/src/Resolver/ChainOrderTypeResolver.php [hunk 1]
```diff
@@ -29,8 +29,8 @@ public function __construct(array $resolvers = []) {
   /**
    * {@inheritdoc}
    */
-  public function addResolver(OrderTypeResolverInterface $resolver) {
-    $this->resolvers[] = $resolver;
+  public function addResolver(OrderTypeResolverInterface $resolver, $priority) {
+    $this->resolvers[$priority] = $resolver;
```
Comment (mglaman) [inline modules/order/src/Resolver/ChainOrderTypeResolver.php:33]: This needs to ksort after add, otherwise priority doesn't mean anything.

**Comment by mglaman** [comment]
@bojanz okay, `$priority = 0` removed. Also added `ksort($this-resolvers)` after adding one, ensures priority is respected. Found that as last minute bug when implementing this patch on project.

File: modules/order/commerce_order.services.yml [hunk 2]
```diff
@@ -8,4 +8,4 @@ services:
     class: Drupal\commerce_order\Resolver\DefaultOrderTypeResolver
     arguments: ['@entity_type.manager']
     tags:
-      - { name: commerce_order.order_type_resolver, priority: 10 }
+      - { name: commerce_order.order_type_resolver, priority: -100 }
```
Comment (mglaman) [inline modules/order/commerce_order.services.yml:11]: Followed the pattern in https://github.com/drupalcommerce/commerce/pull/366 now that I know how priority works O:)

File: tests/src/Unit/Resolver/ChainCountryResolverTest.php [hunk 3]
```diff
@@ -48,9 +48,9 @@ public function testResolver() {
     $third_resolver->expects($this->never())
       ->method('resolve');
 
-    $this->chainCountryResolver->addResolver($first_resolver);
-    $this->chainCountryResolver->addResolver($second_resolver);
-    $this->chainCountryResolver->addResolver($third_resolver);
+    $this->chainCountryResolver->addResolver($first_resolver, 0);
```
Comment (mglaman) [inline tests/src/Unit/Resolver/ChainCountryResolverTest.php]: The priority should be done as `-100`, `100`, and `200`. We should rebuild the container and properly test the priority sorting done when the container is rebuilt.

File: tests/src/Unit/Resolver/ChainLocaleResolverTest.php [hunk 4]
```diff
@@ -48,9 +48,9 @@ public function testResolver() {
     $third_resolver->expects($this->never())
       ->method('resolve');
 
-    $this->chainLocaleResolver->addResolver($first_resolver);
-    $this->chainLocaleResolver->addResolver($second_resolver);
-    $this->chainLocaleResolver->addResolver($third_resolver);
+    $this->chainLocaleResolver->addResolver($first_resolver, 0);
```
Comment (mglaman) [inline tests/src/Unit/Resolver/ChainLocaleResolverTest.php]: The priority should be done as `-100`, `100`, and `200`. We should rebuild the container and properly test the priority sorting done when the container is rebuilt.

**Review by mglaman — COMMENTED** [review]
Need to update tests to properly reflect priority values and how the container is built.

**Comment by mglaman** [comment]
@bojanz want to review this? I've updated the tests to also test priority. I didn't want to mock the container 100%, so I borrowed how it sorts tagged services and mocked that.

**Pull request merged**